File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/ResourceSource.java
package com.ficture7.aasexplorer.model;

import com.ficture7.aasexplorer.client.Client;

import java.net.URI;
import java.util.Date;

import static com.ficture7.aasexplorer.util.ObjectUtil.checkNotNull;

/**
 * Represents a {@link Resource} source.
 *
 * @author FICTURE7
 */
public class ResourceSource implements Source {

    private final Client client;
    private final String name;
    private final Date date;
    private final URI uri;

    /**
     * Constructs a new instance of the {@link ResourceSource} class with the specified
     * {@link Client} instance which acquired the {@link ResourceSource}, resource name,
     * {@link Date} when the source was acquired and {@link URI} pointing to the source.
     *
     * @param client {@link Client} instance which acquired the {@link ResourceSource}.
     * @param name   Resource name.
     * @param date   {@link Date} when the source was acquired.
     * @param uri    {@link URI} pointing to the source.
     * @throws NullPointerException {@code client} is null.
     * @throws NullPointerException {@code name} is null.
     * @throws NullPointerException {@code date} is null.
     * @throws NullPointerException {@code uri} is null.
     */
    public ResourceSource(Client client, String name, Date date, URI uri) {
        this.client = checkNotNull(client, "client");
        this.name = checkNotNull(name, "name");
        this.date = checkNotNull(date, "date");
        this.uri = checkNotNull(uri, "uri");
    }

    /**
     * Returns the {@link Client} instance which acquired the {@link ResourceSource}.
     *
     * @return {@link Client} instance which acquired the {@link ResourceSource}.
     */
    @Override
    public Client client() {
        return client;
    }

    /**
     * Returns the name of the {@link ResourceSource}.
     *
     * @return Name of the {@link ResourceSource}
     */
    public String name() {
        return name;
    }

    /**
     * Returns the {@link Date} when the source was acquired.
     *
     * @return {@link Date} when the source was acquired.
     */
    @Override
    public Date date() {
        return date;
    }

    /**
     * Returns the {@link URI} pointing to the source.
     *
     * @return {@link URI} pointing to the source.
     */
    @Override
    public URI uri() {
        return uri;
    }
}

File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/ResourceSource.kt
package com.ficture7.aasexplorer.model

import com.ficture7.aasexplorer.client.Client

import java.net.URI
import java.util.Date

/**
 * Represents a [Resource] source.
 *
 * @author FICTURE7
 */
class ResourceSource : Source {

    /**
     * Constructs a new instance of the [ResourceSource] class with the specified
     * [Client] instance which acquired the [ResourceSource], resource name,
     * [Date] when the source was acquired and [URI] pointing to the source.
     *
     * @param client [Client] instance which acquired the [ResourceSource].
     * @param name   Resource name.
     * @param date   [Date] when the source was acquired.
     * @param uri    [URI] pointing to the source.
     * @throws NullPointerException `client` is null.
     * @throws NullPointerException `name` is null.
     * @throws NullPointerException `date` is null.
     * @throws NullPointerException `URI` is null.
     */
    constructor(client: Client, name: String, date: Date, uri: URI) {
        this.client = client
        this.name = name
        this.date = date
        this.URI = uri
    }

    /**
     * Returns the [Client] instance which acquired the [ResourceSource].
     *
     * @return [Client] instance which acquired the [ResourceSource].
     */
    override val client: Client
    /**
     * Returns the name of the [ResourceSource].
     *
     * @return Name of the [ResourceSource]
     */
    val name: String

    /**
     * Returns the [Date] when the source was acquired.
     *
     * @return [Date] when the source was acquired.
     */
    override val date: Date

    /**
     * Returns the [URI] pointing to the source.
     *
     * @return [URI] pointing to the source.
     */
    override val URI: URI
}
